File: arrays/max-number-of-k-sum-pairs/max-number-of-k-sum-pairs.java
class Solution {
  public int maxOperations(int[] nums, int k) {
    Arrays.sort(nums);
    int i = 0;
    int j = nums.length - 1;
    int count = 0;

    while (i < j) {
      int sum = nums[i] + nums[j];

      if (sum == k) {
        count++;
        i++;
        j--;
      } else if (sum < k) {
        while (i < j && nums[i] + nums[j] < k) {
          i++;
        }
      } else {
        while (i < j && nums[i] + nums[j] > k) {
          j--;
        }
      }
    }

    return count;
  }
}

class Solution {
  public int maxOperations(int[] nums, int k) {
    Map<Integer, Integer> map = new HashMap<>();
    int count = 0;

    for (int num : nums) {
      map.put(num, map.getOrDefault(num, 0) + 1);
    }

    for (int key : map.keySet()) {
      if (map.containsKey(k - key)) {
        if (k - key == key) {
          count += map.get(key) / 2;
          map.put(key, 0);
        } else {
          count += Math.min(map.get(key), map.get(k - key));
          map.put(k - key, 0);
          map.put(key, 0);
        }
      }
    }

    return count;

  }
}

File: arrays/max-number-of-k-sum-pairs/max-number-of-k-sum-pairs.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
function maxOperations(nums, k) {
  nums.sort((a, b) => a - b);

  let i = 0;
  let j = nums.length - 1;

  let count = 0;

  while (i < j) {
    const sum = nums[i] + nums[j];

    if (sum === k) {
      count++;
      i++;
      j--;
    } else if (sum < k) {
      i++;
      while (i < j && nums[i - 1] === nums[i]) {
        i++;
      }
    } else {
      j--;
      while (i < j && nums[j + 1] === nums[j]) {
        j--;
      }
    }
  }
  return count;
}

function maxOperations(nums, k) {
  const map = {};
  let count = 0;

  for (const num of nums) {
    map[num] = ~~map[num] + 1;
  }

  for (const key in map) {
    if (map[k - key]) {
      // key is a string
      if (k - key == key) {
        count += Math.floor(map[key] / 2);
        map[key] = 0;
      } else {
        count += Math.min(map[key], map[k - key]);
        map[k - key] = 0;
        map[key] = 0;
      }
    }
  }

  return count;
}
